geo/src/algorithm/hausdorff_distance.rs

```diff
@@ -0,0 +1,261 @@
+use crate::algorithm::EuclideanDistance;
+use crate::CoordsIter;
+use crate::GeoFloat;
+use geo_types::*;
+use num_traits::Bounded; // used to have T as generic type in folding
+
+/// Determine the distance between two geometries using the [Hausdorff distance formula].
+///
+/// Hausdorff distance is used to compare two point sets. It measures the maximum euclidean
+/// distance of a point in one set to the nearest point in another set. Hausdorff distance
+/// is often used to measure the amount of mismatch between two sets.
+///
+/// [Hausdorff distance formula]: https://en.wikipedia.org/wiki/Hausdorff_distance
+pub trait HausdorffDistance<T, Rhs = Self> {
+    fn hausdorff_distance(&self, rhs: &Rhs) -> T;
+}
+
+// We can take advantage of the coords_iter() method for all geometries
+// to iterate across all combos of coords (infinum). Simplifies the
+// implementations for all geometries.
+macro_rules! impl_hausdorff_distance_coord_iter {
+    ($from:ident, [$($to:ident),*]) => {
+        $(
+            impl<T> HausdorffDistance<T, $to<T>> for $from<T>
+            where
+                T: GeoFloat
+            {
+                fn hausdorff_distance(&self, geom: &$to<T>) -> T {
+                    // calculate from A -> B
+                    let hd1 = self
+                        .coords_iter()
+                        .map(|c| {
+                            geom
+                            .coords_iter()
+                            .map(|c2| {
+                                c.euclidean_distance(&c2)
+                            })
+                            .fold(<T as Bounded>::max_value(), |accum, val| accum.min(val))
+                        })
+                        .fold(<T as Bounded>::min_value(), |accum, val| accum.max(val));
+
+                    // Calculate from B -> A
+                    let hd2 = geom
+                        .coords_iter()
+                        .map(|c| {
+                            self
+                            .coords_iter()
+                            .map(|c2| {
+                                c.euclidean_distance(&c2)
+                            })
+                            .fold(<T as Bounded>::max_value(), |accum, val| accum.min(val))
+                        })
+                        .fold(<T as Bounded>::min_value(), |accum, val| accum.max(val));
+
+                    // The max of the two
+                    hd1.max(hd2)
+                }
+            }
+        )*
+    };
+}
+
+impl_hausdorff_distance_coord_iter! {
```

Comment:
Thanks for getting comprehensive implementations with the macros. 

I think you could delete a lot of this code if you implemented generically, based on the `CoordsIter` trait.

The impl's would look something like this:

```rust
impl<'a, 'b, T, G1, G2> HausdorffDistance<'a, 'b, T, G2> for G1
where
    T: GeoFloat,
    G1: CoordsIter<'a, Scalar = T>,
    G2: CoordsIter<'b, Scalar = T>,
{
    fn hausdorff_distance(&'a self, rhs: &'b G2) -> T {
       todo!()
    }
}

impl<'a, 'b, T, G> HausdorffDistance<'a, 'b, T, G> for Coord<T>
where
    T: GeoFloat,
    G: CoordsIter<'b, Scalar = T>
{
    fn hausdorff_distance(&'a self, geom: &'b G) -> T {
        todo()!
    }
}
```

... and you'd have to add the lifetimes to the trait declaration. We've done similar things for ConvexHull.

What do you think?

Reply:
I've been trying to get this sorted out for a little while this morning but I'm finding myself in lifetime hell. It seems like there might be a mismatch between lifetimes between `CoordsIter` and `T`. 

I'll check the implementation for ConvexHull and see if i can get it sorted out.

Reply:
Resolved in newest commit.